File: src/test/java/com/outsmart/NQueueImplTest.java
package com.outsmart;

import com.outsmart.measurement.TimedValue;
import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

/**
 * @author Vadim Bobrov
 */
public class NQueueImplTest {

	TimedValue tv1 = new TimedValue(1,1);
	TimedValue tv2 = new TimedValue(2,1);
	TimedValue tv3 = new TimedValue(3,1);
	TimedValue tv3dup = new TimedValue(3,1);
	TimedValue tv4 = new TimedValue(4,1);
	TimedValue tv5 = new TimedValue(5,1);

	@Test
	public void testAll() throws Exception {
		NQueue nQueue = new NQueueImpl();

		assertFalse(nQueue.isFull());

		nQueue.offer(tv1);

		assertSame(tv1, nQueue.get(3));
		assertFalse(nQueue.isFull());

		nQueue.offer(tv2);

		assertSame(tv1, nQueue.get(2));
		assertSame(tv2, nQueue.get(3));
		assertFalse(nQueue.isFull());

		nQueue.offer(tv3);

		assertSame(tv1, nQueue.get(1));
		assertSame(tv2, nQueue.get(2));
		assertSame(tv3, nQueue.get(3));
		assertFalse(nQueue.isFull());

		nQueue.offer(tv4);

		assertSame(tv1, nQueue.get(0));
		assertSame(tv2, nQueue.get(1));
		assertSame(tv3, nQueue.get(2));
		assertSame(tv4, nQueue.get(3));
		assertTrue(nQueue.isFull());

		nQueue.offer(tv5);

		assertSame(tv2, nQueue.get(0));
		assertSame(tv3, nQueue.get(1));
		assertSame(tv4, nQueue.get(2));
		assertSame(tv5, nQueue.get(3));
		assertTrue(nQueue.isFull());
	}

	@Test
	public void testDuplicates() throws Exception {
		NQueue nQueue = new NQueueImpl();

		nQueue.offer(tv1);
		nQueue.offer(tv2);
		nQueue.offer(tv3);
		nQueue.offer(tv3dup);

		assertSame(tv1, nQueue.get(1));
		assertSame(tv2, nQueue.get(2));
		assertSame(tv3, nQueue.get(3));
		assertFalse(nQueue.isFull());
	}

	@Test(expected = AssertionError.class)
	public void testOutOfOrder() throws Exception {
		NQueue nQueue = new NQueueImpl();

		nQueue.offer(tv2);
		nQueue.offer(tv1);
	}


}

File: src/test/scala/com/os/unit/NQueueImplTest.scala
package com.os.unit

import org.scalatest.FlatSpec
import org.scalatest.matchers.ShouldMatchers
import com.os.measurement.TimedValue
import com.os.interpolation.NQueueImpl


/**
 * @author Vadim Bobrov
 */
class NQueueImplTest extends FlatSpec with ShouldMatchers {

	val tv1 = new TimedValue(111, 0.1)
	val tv2 = new TimedValue(222, 0.2)
	val tv3 = new TimedValue(333, 0.3)
	val tv4 = new TimedValue(444, 0.4)
	val tv5 = new TimedValue(555, 0.5)

	val sameAs3 = new TimedValue(333, 0.3)

	"NQueue" should "correctly add one element" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.content should have size (1)
		queue.content.apply(0) should be (tv1)
		queue.get(3) should be (tv1)
		queue.isFull should be (false)
	}

	it should "correctly add two elements" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.offer(tv2)
		queue.content should have size (2)
		queue.content.apply(0) should be (tv1)
		queue.content.apply(1) should be (tv2)
		queue.get(3) should be (tv2)
		queue.get(2) should be (tv1)
		queue.isFull should be (false)
	}

	it should "correctly add three elements" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.offer(tv2)
		queue.offer(tv3)
		queue.content should have size (3)
		queue.content.apply(0) should be (tv1)
		queue.content.apply(1) should be (tv2)
		queue.content.apply(2) should be (tv3)
		queue.get(3) should be (tv3)
		queue.get(2) should be (tv2)
		queue.get(1) should be (tv1)
		queue.isFull should be (false)
	}

	it should "correctly add four elements" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.offer(tv2)
		queue.offer(tv3)
		queue.offer(tv4)
		queue.content should have size (4)
		queue.content.apply(0) should be (tv1)
		queue.content.apply(1) should be (tv2)
		queue.content.apply(2) should be (tv3)
		queue.content.apply(3) should be (tv4)
		queue.get(3) should be (tv4)
		queue.get(2) should be (tv3)
		queue.get(1) should be (tv2)
		queue.get(0) should be (tv1)
		queue.isFull should be (true)
	}

	it should "reject unsorted input" in {
		val queue = new NQueueImpl
		queue.offer(tv2)
		evaluating(queue.offer(tv1)) should produce [IllegalArgumentException]
	}

	it should "ignore nulls" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.offer(null)
		queue.content should have size (1)
		queue.content.apply(0) should be (tv1)
		queue.get(3) should be (tv1)
	}

	it should "ignore duplicates" in {
		val queue = new NQueueImpl
		queue.offer(tv1)
		queue.offer(tv1)
		queue.content should have size (1)
		queue.content.apply(0) should be (tv1)
		queue.get(3) should be (tv1)
	}

}
